src/pudl/metadata/classes.py

```diff
@@ -1691,3 +1692,58 @@ def to_rst(self, top_dir: DirectoryPath, csv_subdir: DirectoryPath, rst_path: st
                 rendered = encoder.to_rst(
                     top_dir=top_dir, csv_subdir=csv_subdir, is_header=header)
                 f.write(rendered)
+
+
+class DataSource(Base):
+    """
+    A data source that has been integrated into PUDL.
+
+    This metadata is used for:
+
+    * Generating PUDL documentation.
+    * Annotating long-term archives of the raw input data on Zenodo.
+    * Defining what data partitions can be processed using PUDL.
+
+    """
+
+    name: SnakeCase
+    title: String = None
+    description: String = None
+    keywords: List[str] = []
+    path: HttpUrl = None
+    contributors: List[Contributor] = []  # Or should this be compiled from Resources?
+    license_raw: License = None
+    license_pudl: License = License.from_id("cc-by-4.0")
+    # concept_doi: Doi = None  # Need to define a Doi type?
+    working_partitions: Dict[SnakeCase, Any] = {}
```

Comment:
I think this would be pretty easy to do with a standalone pydantic model. With our current set of working partitions, I'm thinking something like this:

```py
class WorkingPartitions(Base):
    years: List[int] = None
    states: List[str] = None
    year_month: str = None
```

And you can add some basic validation to this as well

Reply:
I suspect that the one-dimensional partitioning that Dagster supports is a response to some of this fuzziness / complexity. A user can unroll the loops ahead of time or create a generator that provides all the appropriate combinations of year/state/etc. as fully defined work units, and really that process seems external to the code that's going to do the computation on that work unit. And as we've seen it's easy for the simple case of cartesian products to fail to capture the quirks / complexity of defining the valid work units.

I can imagine a more general and better encapsulated solution is to have a method within the `DataSource` class that generates all the valid partitions based on some internal logic and data that's obscured from the outside. In the simple case it's just a list of years, or the cartesian product of two lists (years and states) but it could also deal with weirdness like needing to add PR & AK in a couple of years. This method could take inputs (like `start_year` and `end_year` and `states`), check that the inputs are valid using internal validators, and then return all the valid combinations of those inputs which define the work units to be iterated over or parceled out.